Use `ci-jobs` Utility For `ci.yml` Matrix

ObliviousHarmony opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

This pull request is a follow-up to https://github.com/woocommerce/woocommerce/pull/43345. In that PR I added a new `ci-jobs` command to the `@woocommerce/monorepo-utils` utility. That command replaces the current CI job matrix script with something that is more robust, easier to maintain, and well-tested. In continuing with this line of work I have now made the following changes:

* **`@woocommerce/monorepo-utils` Changes**
  * Reworked the config a little bit to better conform to `ci.yml` requirements.
  * Fixed a couple of bugs that weren't caught previously.
  * Switch build to `webpack` and included the bundled build in the repository. This means that we don't need to run `pnpm install` in order to use the utility anymore. I did this so that we could use it in CI without adding a minute to the runtime for the `ci.yml` workflow.
  * Corrected `bin` packaging and made sure everything works as expected.
* **`ci.yml` Changes**
  * The `ci-jobs` command returns linting and testing as separate jobs. This is deliberate as it makes failures easier to follow when looking at the jobs list.
  * Updated the workflow to take advantage of the new command.
* **`package.json` Changes**
  * The new `config.ci` objects have been filled out according to the needs of each project. These should include information about how to run each command as well as the changes that should trigger each command. 
    * **Please review these thoroughly since people won't notice if a change doesn't trigger a workflow that it should!**

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

Testing this pull request requires creating a PR targeting this branch and making various changes to files to trigger tests. I will do my best to include some examples of this after creation. 

<!-- End testing instructions -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box and supplying data. -->

-   [x] Automatically create a changelog entry from the details below.

<details>

#### Significance

<!-- Choose only one -->

-   [x] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [x] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

#### Comment <!-- If the changes in this pull request don't warrant a changelog entry, you can alternatively supply a comment here. Note that comments are only accepted with a significance of "Patch" -->

This is a CI-only change.
 
</details>

ObliviousHarmony committed:
Globs Instead Of Regexes For Change Config

Realistically speaking, all of our other tools use a glob
pattern when dealing with paths. For the sake of consistency,
let's make sure to do the same here.

ObliviousHarmony committed:
Made Graph Test More Robust

ObliviousHarmony committed:
Added Lint Configs

ObliviousHarmony committed:
Prevented Duplicate Job Creation

When I wrote the job processor initially I forgot to track
the visited job nodes. This led to nodes that were
referenced multiple times in the graph having
duplicate jobs created when matching.

ObliviousHarmony committed:
Expanded CI CLI Output

Let's make sure that the output looks nice both in CI
and in the terminal.

ObliviousHarmony committed:
Added Test Config

ObliviousHarmony committed:
Revised CI Matrix Generation

ObliviousHarmony committed:
Bundled Monorepo Utils

Rather than building the utils we will be including them
in version control. This means we won't need to build
the utils package before using it.

ObliviousHarmony committed:
CI Trigger

ObliviousHarmony committed:
Removed Unnecessary Install & Build

ObliviousHarmony committed:
Ensure Font Line Endings

ObliviousHarmony committed:
Logging

ObliviousHarmony committed:
Revised GitHub Output

ObliviousHarmony committed:
Added Linting Job

ObliviousHarmony committed:
Fix Matrix Input

ObliviousHarmony committed:
Added Test Jobs

ObliviousHarmony committed:
Logging

ObliviousHarmony committed:
Logging

ObliviousHarmony committed:
Testing

ObliviousHarmony committed:
Logging

ObliviousHarmony committed:
Changed Test Env Structure

ObliviousHarmony committed:
Fixed Test Env Creation Check

ObliviousHarmony committed:
Fixed Test Env Var Mapping

Since the environment variable binding runs before the
if statement we need to have something to bind. This
means we need an empty object present
whether or not the environment is made.

ObliviousHarmony committed:
Fixed CI Test Breakage

ObliviousHarmony committed:
Testing

ObliviousHarmony committed:
Removed Testing Addition

ObliviousHarmony committed:
Fixed WooCommerce Test Command

ObliviousHarmony committed:
Test Trigger

ObliviousHarmony committed:
Removed Test Trigger

ObliviousHarmony committed:
Added Command Var Replacement

ObliviousHarmony committed:
Linting Fixes

Commit:
Add changefile(s) from automation for the following project(s): @woocommerce/tracks, @woocommerce/product-editor, @woocommerce/onboarding, @woocommerce/number, @woocommerce/notices, @woocommerce/navigation, @woocommerce/internal-js-tests, @woocommerce/integrate-plugin, @woocommerce/expression-evaluation, @woocommerce/explat, @woocommerce/experimental, @woocommerce/eslint-plugin, @woocommerce/dependency-extraction-webpack-plugin, @woocommerce/date, @woocommerce/data, @woocommerce/customer-effort-score, @woocommerce/currency, @woocommerce/csv-export, @woocommerce/components, @woocommerce/block-templates, @woocommerce/api, @woocommerce/ai, @woocommerce/admin-layout, @woocommerce/admin-e2e-tests, woocommerce-blocks, woocommerce

ObliviousHarmony commented:
Yes @jonathansadowski, it's because I renamed the job.

rrennick commented:
> just to verify, the current expected checks not finishing is expected for now, as after this is merged, we'd update all of those separate expected checks and replace them with a single "Evaluate Project Job Statuses" expected check, is that correct?

This is the list of jobs picked up by this PR with `pnpm utils ci-jobs trunk`. My trunk might be slightly out of date.

```
-  @woocommerce/monorepo-utils - lint
-  @woocommerce/data - lint
-  @woocommerce/product-editor - lint
-  @woocommerce/admin-library - lint
-  @woocommerce/block-library - lint
-  @woocommerce/plugin-woocommerce - lint:php:changes:branch trunk
Test Jobs
-  @woocommerce/monorepo-utils - JavaScript
-  @woocommerce/data - JavaScript
-  @woocommerce/components - JavaScript
-  @woocommerce/product-editor - JavaScript
-  @woocommerce/admin-library - JavaScript
-  @woocommerce/block-library - JavaScript
-  @woocommerce/plugin-woocommerce - PHP
-  @woocommerce/plugin-woocommerce - PHP 8.0
-  @woocommerce/plugin-woocommerce - PHP WP: nightly
-  @woocommerce/plugin-woocommerce - PHP WP: latest - 1
-  @woocommerce/plugin-woocommerce - PHP WP: latest - 2
```

rrennick reviewed (commented):
@jonathansadowski This looks good to me but I'll leave it for you to approve.

rrennick commented:
When I tried to switch way from this branch I got the following error:

```
error: Your local changes to the following files would be overwritten by checkout:
	tools/monorepo-utils/fonts/Standard.flf
```

Should this file be added to the PR or to `.gitignore`?

jonathansadowski commented on packages/js/api-core-tests/package.json:
```diff
@@ -49,5 +49,18 @@
+			"lint": {
+				"command": "lint",
+				"changes": [
+					"data/**/*.js",
+					"endpoints/**/*.js",
+					"tests/**/*.js",
+					"utils/**/*.js"
```
Currently these directories only contain `.js` files, but the lint command itself contains the args `--ext=js,ts,tsx`, should we be more inclusive here?

ObliviousHarmony replied:
Good catch, I must have decided to check the `--ext` option later on in the process of converting these.

ObliviousHarmony committed:
Added Font Line Ending Replacement

ObliviousHarmony commented:
> Should this file be added to the PR or to .gitignore?

No @rrennick, this file is necessary in order for the `@woocommerce/monorepo-utils` utility to work without being built. We've got the `.gitattributes` to automatically set the line ending, however, it seems that `git` gives some messaging when it needs to do this conversion depending on your configuration. I've added a step to the build for the package that automatically sets the line endings to `\n` and that should solve any issues relating to this.

ObliviousHarmony committed:
Fixed api-core-tests Lint Changes

ObliviousHarmony committed:
Use Package File For Dependencies

Since we can't use `pnpm list` in CI unless we do
a `pnpm install` we will need to pull the dependency
information directly from the `package.json` files.

ObliviousHarmony committed:
Simplied Dependency Checking

jonathansadowski reviewed (approved):
I've tested this with several changes that trigger both lint and test jobs. I've also reviewed all of the `package.json` changes, and they look correct to me. I didn't both trigger a test job and lint job for each project, but I've tested out the various scenarios, and this looks like it's good to go to me.

Nice work!

ObliviousHarmony committed:
Added Comment About Cascade Keys
